### PR TITLE
Contracts performance measurement tool proposal

### PR DESCRIPTION
**Pull request by smiasojed**
#### Project Abstract

The key objective of this grant is to build a tool which allows to compare contracts performance and to track its regressions.
Tool is integrated with CI and triggered by a release of pallet-evm or pallet-contracts based parachain. It measures the contract performance and allows for its visualization.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [x] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@smiasojed:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — DISMISSED** [review]
Thanks for the application. It looks good to me, and I'm happy to go ahead with it as well as share it with the rest of the team.

File: applications/contracts-tool.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+- **Estimated duration:** 26 days
+- **FTE:**  0.5
+- **Costs:** 6500 EUR
+
+| Number | Deliverable | Specification |
+| -----: | ----------- | ------------- |
+| 1. | License | GPLv3 |
+| 2. | Documentation | We will provide both **inline documentation** of the code and a basic **tutorial** that explains how a user can spin up test net and run contracts with transactions. |
+| 3. | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| 4. | Updated evm contracts | We will update tool with support for newest Moonbeam parachain. |
```
Comment (semuelle) [inline applications/contracts-tool.md:153]: What do you mean by "updated EVM contracts"? And is there anything specific to Moonbeam about any of the deliverables that wouldn't also work for other EVM-compatible parachains, or why are you singling out Moonbeam here?

**Review by semuelle — COMMENTED** [review]
Thanks for the application, @smiasojed. Generally happy with the application, just one question below.

**Comment by keeganquigley** [comment]
Thanks for the application @smiasojed it looks good and it's cool to see support for Solang as well. Just a couple of minor suggestions:

- To match our [template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md), mandatory deliverables should start with 0a. - 0d. instead of 1. Then you can start with 1 for the core deliverables (i.e. 1. Updated EVM Contracts). Can you update this?
- Typically, the milestone estimated durations are written in weeks instead of days. This is just a convention but helps to avoid confusion later on when keeping track of progress.

Thanks!

**Comment by keeganquigley** [comment]
Thanks @smiasojed I see you changed the first milestone but not the second, this would go for that one as well.

**Comment by smiasojed** [comment]
> Thanks @smiasojed I see you changed the first milestone but not the second, this would go for that one as well.

I missed it, I am sorry. Now should be better

**Review by keeganquigley — APPROVED** [review]
Thanks! LGTM

**Review by BenWhiteJam — APPROVED** [review]

**Comment by smiasojed** [comment]
We are planning to close the 1st milestone until 09.07.2023

**Comment by keeganquigley** [comment]
Hi @smiasojed how is the 2nd milestone coming along?

**Comment by smiasojed** [comment]
Hi @keeganquigley, We've encountered a delay, and we are currently working on updating smart-bench dependencies to the newest versions. As for phase 2, half of it has been completed, and we estimate that we should be ready by the end of the year.

**Comment by keeganquigley** [comment]
Thanks @smiasojed in this case could you please file an amendment to reflect the changes and extend the timeline? You can amend the original application. Thanks!

**Comment by keeganquigley** [comment]
Hi @smiasojed are you still planning to submit an amendment?

**Comment by semuelle** [comment]
Hi @smiasojed. How is M2 progressing?

**Comment by smiasojed** [comment]
> Hi @smiasojed. How is M2 progressing?

Hi @semuelle, we have all the code in pull requests, and we are waiting for approvals from the Parity team's side.
Once it is accepted I will submit M2.

**Comment by keeganquigley** [comment]
Hi @smiasojed just checking in for an update on M2?

**Comment by keeganquigley** [comment]
Hi @smiasojed since the first PR has been merged feel free to go ahead and submit the delivery in the [milestone delivery repo](https://github.com/w3f/Grant-Milestone-Delivery/pulls). That way we can start evaluating and it will take the pressure off of the timeline. Besides I'm not seeing a hard requirement in the milestone deliverables that says they need to be merged into the codebase, so we'll probably be able to go ahead and pass it as is.

**Comment by smiasojed** [comment]
Hi @keeganquigley, thanks! I will do it next week.

**Pull request merged**